File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    
    void rotateArray(vector<int> &v, int k){
        int n = v.size();
        if(k % n == 0){
            return;
        }
        if(k > n){
            k = k % n;
        }
        reverse(v.begin(), v.end());
        reverse(v.begin(), v.begin() + k);
        reverse(v.begin() + k, v.end());
    }

    void rotate(vector<int>& nums, int k) {
        rotateArray(nums, k);
    }
};

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        k = k % n

        # now reverse left, right, full
        
        # left -> [0, n - k - 1] elements
        nums[:n-k] = nums[:n-k][::-1]
        print(nums)
        # right -> [n - k, n - 1] elements
        nums[n-k:] = nums[n-k:][::-1]
        print(nums)
        # full reverse
        nums[:] = nums[::-1]
        


        